docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,259 @@
+# Creation and Distribution of Secrets
+
+## Status
+
+** proposed **
+
+## Context
+
+This ADR seeks to clarify and prioritize the secret handling approach taken by EdgeX.
+
+EdgeX microservices need a number of secrets to be created and distributed
+in order to create a functional, secure system.
+Among these secrets are:
+
+- Privileged administrator passwords (such as a database superuser password)
+- Service account passwords (e.g. for a database)
+- PKI certificates
+
+There is a lack of consistency on how secrets are created and distributed to EdgeX microservices,
+and when developers need to add new components to the system,
+it is unclear on what the preferred approach should be.
+
+This document assumes a threat model wherein the EdgeX services are sandboxed
+(such as in a snap or a container) and the host system is trusted,
+and all services running in a single snap share a trust boundary.
+
+### Current practices survey
+
+A survey on the existing EdgeX secrets reveals the following appoaches.
+
+A designation of "compliant" means that the current implementation
+is aligned with the recommended practices documented in the next section.
+A designation of "non-compliant" means that the current implementation
+uses an implemention mechanism outside of the recommended practices documented in the next section.
+A "non-compliant" implementation is a candidate for refactoring
+to bring the implementation into conformance with the recommended practices.
+
+- PKI private keys
+  * Docker: PKI generated by standalone utility every boot. Distribution via secrets volume. (Compliant.)
+  * Snaps: PKI generated by standalone utility every boot. Deployed to `$SNAP_DATA/secrets`. (Compliant.)
+
+- Secret store master password
+  * Docker: Distribution via persistent docker volume. (Non-compliant.)
+  * Snaps: Stored in `$SNAP_DATA/config/security-secrets-setup/res`. (Non-compliant.)
+
+- Secret store per-service authentication tokens
+  * Docker: Distribution via secrets volume, generated every boot. (Compliant.)
+  * Snaps: Distribution via secrets volume, generated every boot. (Compliant.)
+
+- Postgres superuser password
+  * Docker: Hard-coded into docker-compose file, checked in to source control. (Non-compliant.)
+  * Snaps: Deployed to `$SNAP_DATA/config/postgres/kongpw`. (Non-compliant.) 
+
+- MongoDB service account passwords
```

Comment:
Should this be deprecated since we are using Redis from now on by default?

Reply:
> Should this be deprecated since we are using Redis from now on by default?

Hard to tell what lines were highlighted, but:

Postgres: Kong only works with Postgres and Cassandra not Redis so Postgres has to stay.
MongoDB: We reverted the PR to remove PW generation for MongoDB just in case someone wants to keep using it, so I think this should stay as well.